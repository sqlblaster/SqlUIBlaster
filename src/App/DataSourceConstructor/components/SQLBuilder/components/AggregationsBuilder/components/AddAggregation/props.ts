import { actionCreators, PickActionCreators } from 'src/components/query-builder/store/action-creators';
import { PickStates } from 'src/components/query-builder/store/models/State';
import { PopoverManagerProps } from '../../../common/WithPopoverManagement';

export type AddAggregationOwnProps = PopoverManagerProps;

export type AddAggregationStateProps = PickStates<'aggregations'>;

export type AddAggregationDispatchProps = PickActionCreators<'addAggregation'>;

export const AAMapDispatchToProps: AddAggregationDispatchProps = {
  addAggregation: actionCreators.addAggregation
};

export type AddAggregationProps = AddAggregationStateProps & AddAggregationDispatchProps;

export type AddAggregationViewProps = AddAggregationOwnProps &
  AddAggregationStateProps &
  AddAggregationDispatchProps;
