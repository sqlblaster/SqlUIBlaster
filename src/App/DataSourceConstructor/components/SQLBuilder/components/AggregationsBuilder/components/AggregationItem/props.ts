import { ClassesProp } from 'src/components/query-builder/App/utils/classes-prop';
import { actionCreators, PickActionCreators } from 'src/components/query-builder/store/action-creators';
import { PopoverManagerProps } from '../../../common/WithPopoverManagement';
import { Aggregation } from './model';
import { AggregationItemClassKeys } from './styles';

export interface AggregationItemOwnProps {
  aggregation: Aggregation;
}

export type AggregationItemDispatchProps = PickActionCreators<
  'removeAggregation' | 'updateAggregation'
>;

export const AIMapDispatchToProps: AggregationItemDispatchProps = {
  removeAggregation: actionCreators.removeAggregation,
  updateAggregation: actionCreators.updateAggregation
};

export type AggregationItemProps = AggregationItemOwnProps &
  AggregationItemDispatchProps;

export type AggregationItemViewProps = {
  handleAggregationRemoval: (id: Aggregation['id']) => () => void;
} & AggregationItemOwnProps &
  PopoverManagerProps &
  PickActionCreators<'updateAggregation'> &
  ClassesProp<AggregationItemClassKeys>;
