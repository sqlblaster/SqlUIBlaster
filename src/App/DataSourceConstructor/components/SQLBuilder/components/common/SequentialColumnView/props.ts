import { ClassesProp } from 'src/components/query-builder/App/utils/classes-prop';
import { ColumnBranchFields } from '../ColumnSelector/ColumnItem/SelectedColumn.models';
import { SequentialColumnViewClassKeys } from './styles';

export type SequentialColumnViewProps = ColumnBranchFields &
  SequentialColumnViewConfigProps &
  ClassesProp<SequentialColumnViewClassKeys>;

export class SequentialColumnViewConfigProps {
  public isForeignColumn?: boolean;
  public prefix?: string = '';
  public noSpaceAfterPrefix?: boolean = false;
  public suffix?: string = '';
  public noSpaceBeforeSuffix?: boolean = false;
}
