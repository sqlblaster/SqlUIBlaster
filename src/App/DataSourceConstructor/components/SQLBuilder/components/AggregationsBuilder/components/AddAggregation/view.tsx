import * as React from 'react';
import { AddEntity } from '../../../common/AddEntity';
import { MutateAggregationPopover } from '../MutateAggregationPopover';
import { AddAggregationViewProps } from './props';

export const addAggregationText = 'Add an aggregation';

// FIXME: make padding clickable too
export class AddAggregation extends React.Component<AddAggregationViewProps, any> {
  private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

  constructor(props: AddAggregationViewProps) {
    super(props);

    this.popoverLauncher = React.createRef<HTMLDivElement>();
  }

  public render() {
    const {
      aggregations,
      addAggregation,
      isOpen,
      handleClose,
      handleOpen
    } = this.props;

    return (
      <>
        <AddEntity
          showText={!(aggregations && aggregations.length)}
          onClick={handleOpen}
          reference={this.popoverLauncher}
        >
          {addAggregationText}
        </AddEntity>
        { isOpen && (// TODO: resolve this problem another way
          <MutateAggregationPopover
            action={addAggregation}
            isOpen={isOpen}
            anchorEl={this.popoverLauncher.current}
            onClose={handleClose}
          />
        ) }
      </>
    );
  }
}
