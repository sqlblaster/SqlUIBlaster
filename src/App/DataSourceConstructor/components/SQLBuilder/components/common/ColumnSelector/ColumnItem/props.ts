import { CollapseProps } from '@material-ui/core/Collapse';
import { Column } from 'src/components/query-builder/App/DataSourceConstructor/schemas';
import { TableColumnsDelegatedProps } from '../props';
import { ColumnBranch } from './SelectedColumn.models';

export type ColumnItemProps = Column &
TableColumnsDelegatedProps &
ColumnItemDelegatedProp & {
  columnBranch: ColumnBranch;
};

export interface IColumnItemHandlers {
  handleColumnClick: () => void;
  handleCollapsedStateChange: (event: React.MouseEvent) => void;
  handleCollapseEnter: CollapseProps['onEntered'];
  handleCollapseExit: CollapseProps['onExited'];
}

export type ColumnItemViewProps = ColumnItemProps & {
  collapsed: boolean;
} & IColumnItemHandlers;

export interface ColumnItemDelegatedProp {
  level: number;
}
