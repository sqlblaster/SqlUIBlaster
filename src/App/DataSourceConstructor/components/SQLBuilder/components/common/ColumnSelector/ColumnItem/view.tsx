import {
  Collapse,
  IconButton,
  ListItem,
  ListItemIcon,
  ListItemText,
  withStyles
} from '@material-ui/core';
import ExpandLess from '@material-ui/icons/ExpandLess';
import ExpandMore from '@material-ui/icons/ExpandMore';
import * as React from 'react';
import { TableColumnsDelegatedProps } from '../props';
import { TableColumns } from '../view';
import { ColumnItemViewProps } from './props';
import { TypeIcon } from './TypeIcon';
import { calcPaddingAccordingTo, getNewTableBranch } from './utils';

export const ColumnItem = withStyles({})(
  class extends React.Component<ColumnItemViewProps> {
    public render() {
      const {
        name,
        type,
        foreignTableName,
        foreignModelName,
        columnBranch,
        level,
        tables,
        popoverPositionUpdater,
        onColumnSelected,
        canShowColumnBranch,
        collapsed,
        handleColumnClick,
        handleCollapsedStateChange,
        handleCollapseEnter,
        handleCollapseExit
      } = this.props;

      const delegatedProps: TableColumnsDelegatedProps = {
        tables,
        popoverPositionUpdater,
        onColumnSelected,
        canShowColumnBranch
      };

      const disabledForeignColumn = canShowColumnBranch
        ? !canShowColumnBranch(columnBranch) && !!foreignTableName
        : false;

      return (
        <>
          <ListItem
            {...(disabledForeignColumn
              ? {}
              : { button: true, onClick: handleColumnClick })}
            style={{
              paddingLeft: calcPaddingAccordingTo(level),
              ...(disabledForeignColumn ? { cursor: 'default' } : {})
            }}
          >
            <ListItemIcon>
              <TypeIcon type={foreignTableName ? 'foreign-column' : type} />
            </ListItemIcon>
            <ListItemText inset={true} primary={name} />
            {foreignTableName && (
              <>
                {collapsed ? (
                  <IconButton
                    onClick={handleCollapsedStateChange}
                    style={{ padding: 0 }}
                  >
                    <ExpandMore />
                  </IconButton>
                ) : (
                  <IconButton
                    onClick={handleCollapsedStateChange}
                    style={{ padding: 0 }}
                  >
                    <ExpandLess />
                  </IconButton>
                )}
              </>
            )}
          </ListItem>

          {foreignTableName && (
            <Collapse
              in={!collapsed}
              timeout='auto'
              unmountOnExit={true}
              onEntered={handleCollapseEnter}
              onExited={handleCollapseExit}
            >
              <TableColumns
                {...delegatedProps}
                level={level + 1}
                tableName={foreignTableName}
                tableBranch={getNewTableBranch(
                  columnBranch,
                  foreignTableName,
                  foreignModelName
                )}
              />
            </Collapse>
          )}
        </>
      );
    }
  }
);
