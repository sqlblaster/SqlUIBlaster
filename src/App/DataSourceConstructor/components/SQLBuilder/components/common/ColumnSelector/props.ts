import { ClassesProp } from 'src/components/query-builder/App/utils/classes-prop';
import { PickStates } from 'src/components/query-builder/store/models/State';
import { ColumnItemDelegatedProp } from './ColumnItem/props';
import { ColumnBranch } from './ColumnItem/SelectedColumn.models';
import { TableColumnsClassKeys } from './styles';

export interface ColumnSelectorOwnProps {
  popoverPositionUpdater: (() => void) | undefined;
  canShowColumnBranch?: (columnBranch: ColumnBranch) => boolean;
  onColumnSelected: (column: ColumnBranch) => void;
}

export type ColumnSelectorStateProps = PickStates<'tables'> & {
  tableName: string | null;
  modelName: string | null;
};

export type ColumnSelectorProps = ColumnSelectorStateProps &
  ColumnSelectorOwnProps;

export type TableColumnsDelegatedProps = ColumnSelectorOwnProps &
  Pick<ColumnSelectorStateProps, 'tables'>;

export type TableColumnsProps = {
  /**
   * This is columnBranch, only with tables' values declared
   */
  tableBranch: ColumnBranch;
} & Pick<ColumnSelectorStateProps, 'tableName'> &
  TableColumnsDelegatedProps &
  ColumnItemDelegatedProp &
  ClassesProp<TableColumnsClassKeys>;
