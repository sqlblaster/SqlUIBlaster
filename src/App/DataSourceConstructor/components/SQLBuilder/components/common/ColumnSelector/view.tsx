import { List, Typography, withStyles } from '@material-ui/core';
import * as React from 'react';
import { ColumnItem } from './ColumnItem';
import { ColumnItemDelegatedProp } from './ColumnItem/props';
import { TableColumnsDelegatedProps, TableColumnsProps } from './props';
import { tableColumnsStyles } from './styles';
import { filterColumns, getNewColumnBranch, sortInOrderForeignColumnsBelow } from './utils';

export const TableColumns = withStyles(tableColumnsStyles)((props => {
  const {
    tableName,
    tables,
    tableBranch,
    level,
    popoverPositionUpdater,
    classes,
    onColumnSelected,
    canShowColumnBranch
  } = props;

  const table = tables.find(t => !!tableName && t.name === tableName.toLowerCase());
  let columns = table && table.columns;

  columns &&
    canShowColumnBranch &&
    (columns = filterColumns(columns, canShowColumnBranch, tableBranch));

  columns && (columns = sortInOrderForeignColumnsBelow(columns));

  const delegatedProps: ColumnItemDelegatedProp & TableColumnsDelegatedProps = {
    level,
    tables,
    popoverPositionUpdater,
    onColumnSelected,
    canShowColumnBranch
  };

  return (
    <div className={classes.root}>
      {columns && columns.length ? (
        <List component='nav' disablePadding={true}>
          {columns.map(column => (
            <ColumnItem
              key={column.name}
              {...column}
              foreignTableName={column.foreignTableName}
              {...delegatedProps}
              columnBranch={getNewColumnBranch(tableBranch, column)}
            />
          ))}
        </List>
      ) : (
        <Typography className={classes.noColumns} data-testid='no-columns'>
          No columns available...
        </Typography>
      )}
    </div>
  );
}) as React.FC<TableColumnsProps>);
