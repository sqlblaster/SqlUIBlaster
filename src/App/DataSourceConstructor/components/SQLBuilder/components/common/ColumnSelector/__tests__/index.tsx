import * as React from 'react';
import { Table } from 'src/App/DataSourceConstructor/schemas';
import { getConfiguredRender } from 'src/App/DataSourceConstructor/test-utils/configured-render';
import { WithReduxProvider } from 'src/App/DataSourceConstructor/test-utils/contexts-provider/with-ReduxProvider';
import { getCertainCalledProps } from 'src/App/DataSourceConstructor/test-utils/react-utils';
import { getStore } from 'src/App/DataSourceConstructor/test-utils/store/getStore';
import { ColumnSelector } from '..';
import { ColumnSelectorOwnProps, TableColumnsProps } from '../props';
import { TableColumns } from '../view';

jest.mock('../view', () => ({
  TableColumns: jest.fn(() => null)
}));

const render = getConfiguredRender<ColumnSelectorOwnProps>(
  <ColumnSelector
    onColumnSelected={jest.fn()}
    popoverPositionUpdater={undefined}
  />
);

describe('ColumnSelector', () => {
  const TableColumnsMock = TableColumns as jest.Mock;

  afterEach(() => {
    TableColumnsMock.mockClear();
  });

  it('should not render TableColumns component if there is no picked table', () => {
    render(
      {},
      {
        contextProviderApplicators: [
          WithReduxProvider(
            getStore({
              selectedTable: null
            })
          )
        ]
      }
    );

    expect(TableColumnsMock).not.toHaveBeenCalled();
  });

  it(`should pass initial level value, root table branch,
    and redux states to TableColumns component`, () => {
    const tableName = 'Users';

    render(
      {},
      {
        contextProviderApplicators: [
          WithReduxProvider(
            getStore({
              selectedTable: new Table({ name: tableName })
            })
          )
        ]
      }
    );

    const certainProps = getCertainCalledProps<TableColumnsProps>(
      TableColumnsMock,
      ['level', 'tableBranch']
    );

    expect(certainProps).toMatchSnapshot();
  });
});
