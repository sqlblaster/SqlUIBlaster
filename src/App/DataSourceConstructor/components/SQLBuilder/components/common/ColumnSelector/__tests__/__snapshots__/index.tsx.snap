// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnSelector should pass initial level value, root table branch,
    and redux states to TableColumns component 1`] = `
Object {
  "level": 0,
  "tableBranch": Object {
    "_foreignTableName": undefined,
    "_lastColumn": undefined,
    "columnName": "",
    "foreignColumn": undefined,
    "foreignModelName": undefined,
    "foreignTableName": undefined,
    "modelName": "Users",
    "tableName": "users",
    "type": undefined,
  },
}
`;
